lib/Reader/readerir.cpp

```diff
@@ -2422,20 +2451,40 @@ void GenIR::createArrayOfReferenceType() {
   Fields.push_back(VtablePtrTy);
 
   // Fill in the rest.
+
+  // In case ElementTy is null, setup ObjectClassHandle for System.Object.
+  // It might make sense for this to create System.__Canon[]* but there's
+  // no builtin for the canon type. So we create System.Object[] *.
+  // If ElementTy is not null then ObjectClassHandle and CORINFO_TYPE_CLASS
+  // will be ignored.
   CORINFO_CLASS_HANDLE ObjectClassHandle =
       getBuiltinClass(CorInfoClassId::CLASSID_SYSTEM_OBJECT);
   const uint32_t ArrayRank = 1;
   const bool IsZeroLowerBoundSDArray = true;
   const bool GetElementAggregateFields = true;
-  addArrayFields(Fields, IsZeroLowerBoundSDArray, ArrayRank, CORINFO_TYPE_CLASS,
-                 ObjectClassHandle, GetElementAggregateFields);
+  addArrayFields(Fields, IsZeroLowerBoundSDArray, ArrayRank, ElementTy, 
+                 CORINFO_TYPE_CLASS, ObjectClassHandle, GetElementAggregateFields);
 
   // Install fields and give this a recognizable name.
   StructTy->setBody(Fields, true /* isPacked */);
-  StructTy->setName("__ArrayOfReferences");
+  std::string TypeName;
+
+  // We use a raw_string_ostream as a handly way to construct
+  // the name of the array type. This lets us use the
+  // Type::print method to get the element type name.
+  raw_string_ostream StringStream(TypeName);
+  StringStream << "__ArrayOf__";
```

Comment:
We should try to have names that look similar to what we get for arrays from getClassType. I think they look like object[].